File: Solutions/PE012_highly_divisible_triangular_number/python/highly_divisible_triangular_number.py
#!/bin/python3

# https://www.hackerrank.com/contests/projecteuler/challenges/euler012


def sieve_of_eratosthenes(n):
    """
    Return list of primes less than n
    """
    res = [2]
    i = 3
    marked = set()
    while i <= n**.5:
        if i not in marked:
            res.append(i)
            j = 0
            while j <= n/i:
                marked.add(i + j*i)
                j += 1
        i += 2
    while i <= n:
        if i not in marked:
            res.append(i)
        i += 2
    return res


def count_divisors(n, primes):
    """
    counts divisors using a table of primes
    """
    # primes = Sieve_of_Eratosthenes(1000000)
    i = 0
    divisors = 1
    while primes[i] <= n:
        p = primes[i]
        count = 0
        while n % p == 0:
            count += 1
            n //= p
        if count > 0:
            divisors *= count + 1
        i += 1
    return divisors


def triangle_number_with_more_than_n_divisors(n):
    i = 1
    triangle = 1
    divisors = 1
    m = 1
    table = dict()
    table[0] = triangle
    primes = sieve_of_eratosthenes(1000000)
    while divisors <= n:
        i += 1
        triangle += i
        divisors = count_divisors(triangle, primes)
        if divisors > m:
            for j in range(m, divisors):
                table[j] = triangle
            m = divisors
    return table


if __name__ == '__main__':
    table = triangle_number_with_more_than_n_divisors(500)
    print(table[500])
    # table = triangle_number_with_more_than_n_divisors(1000)
    # t = int(input())
    # for _ in range(t):
    #     n = int(input())
    #     print(table[n])

File: Solutions/PE012_highly_divisible_triangular_number/cpp/HighlyDivisibleTriangularNumber.cpp
#include <iostream>
#include <math.h>
#include <vector>
#include <set>
#include <map>

//https://www.hackerrank.com/contests/projecteuler/challenges/euler012

using namespace std;

vector<long> fastSieveOfEratosthenes(long n) {
	vector<bool> sieve((int)n / 2, true);
	for (int i = 3; i < sqrt(n) + 1; i += 2) {
		if (sieve[i / 2]) {
			for (int j = i * i / 2; j < sieve.size(); j += i) {
				sieve[j] = false;
			}
		}
	}
	vector<long> res;
	res.push_back(2);
	for (long i = 1; i < n / 2; i++) {
		if (sieve[i])
			res.push_back(2 * i + 1);
	}
	return res;
}

int countDivisors(int n, vector<long> primes) {
	//counts divisors using a vector of primes
	int i = 0;
	int res = 1;
	while (primes[i] <= n) {
		long p = primes[i];
		int count = 0;
		while (n%p == 0) {
			count++;
			n /= p;
		}
		if (count > 0)
			res *= count + 1;
		i++;
	}
	return res;
}

map<int, long> triangleNumberWithNDivisors(long n) {
	int i = 1;
	long triangle = 1;
	int divisors = 1;
	int m = 1;
	map<int, long> table;
	table[0] = triangle;
	vector<long> primes = fastSieveOfEratosthenes(100000);
	while (divisors <= n) {
		i++;
		triangle += i;
		divisors = countDivisors(triangle, primes);
		if (divisors > m) {
			for (int j = m; j < divisors; j++)
				table[j] = triangle;
			m = divisors;
		}
	}
	return table;
}

int main() {
	map<int, long > table = triangleNumberWithNDivisors(1000);
	int t;
	cin >> t;
	for (int i = 0; i < t; i++) {
		int n;
		cin >> n;
		cout << table[n] << endl;
	}

}
